driver-core/src/main/com/mongodb/connection/ScramShaAuthenticator.java

```diff
@@ -190,78 +201,67 @@ public void dispose() {
         private byte[] decodeUTF8(final String str) throws SaslException {
             try {
                 return str.getBytes("UTF-8");
-            }
-            catch (UnsupportedEncodingException e) {
+            } catch (UnsupportedEncodingException e) {
                 throw new SaslException("UTF-8 is not a supported encoding.", e);
             }
         }
 
+        private String encodeBase64(final String str) throws SaslException {
+            return Base64.encode(decodeUTF8(str));
+        }
+
         private String encodeBase64(final byte[] bytes) {
             return Base64.encode(bytes);
         }
 
         private String encodeUTF8(final byte[] bytes) throws SaslException {
             try {
                 return new String(bytes, "UTF-8");
-            }
-            catch (UnsupportedEncodingException e) {
+            } catch (UnsupportedEncodingException e) {
                 throw new SaslException("UTF-8 is not a supported encoding.", e);
             }
         }
 
         private byte[] h(final byte[] data) throws SaslException {
             try {
-                return MessageDigest.getInstance("SHA-1").digest(data);
-            }
-            catch (NoSuchAlgorithmException e) {
-                throw new SaslException("SHA-1 could not be found.", e);
+                return MessageDigest.getInstance(hAlgorithm).digest(data);
+            } catch (NoSuchAlgorithmException e) {
+                throw new SaslException(format("Algorithm for '%s' could not be found.", hAlgorithm), e);
             }
         }
 
-        private byte[] hi(final String password, final byte[] salt, final int iterations) throws SaslException {
-            PBEKeySpec spec = new PBEKeySpec(password.toCharArray(), salt, iterations, 20 * 8 /* 20 bytes */);
-
-            SecretKeyFactory keyFactory;
-            try {
-                keyFactory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA1");
-            }
-            catch (NoSuchAlgorithmException e) {
-                throw new SaslException("Unable to find PBKDF2WithHmacSHA1.", e);
-            }
-
+        private byte[] hi(final byte[] password, final byte[] salt, final int iterations) throws SaslException {
             try {
-                SecretKey secretKey = keyFactory.generateSecret(spec);
-                // Workaround for https://bugs.openjdk.java.net/browse/JDK-8191177, as suggested in
-                // https://bugs.openjdk.java.net/browse/JDK-8055183
-                //noinspection SynchronizationOnLocalVariableOrMethodParameter
-                synchronized (secretKey) {
-                    return secretKey.getEncoded();
+                SecretKeySpec key = new SecretKeySpec(password, hmacAlgorithm);
```

Comment:
For the record, I realized why I failed in my attempt to prove that PBKDF2WithHmacSHA1/PBKDF2WithHmacSHA256 SecretKeyFactory was identical to this code: I had neglected to properly set the keyLength in the PBEKeySpec to 32 * 8 for PBKDF2WithHmacSHA256.

Reply:
I said as much in response to the previous PR where you asked - glad you are finally reassured that it works to spec.

Reply:
Aghh, I somehow missed that comment since the PR was closed soon after.